### PR TITLE
Add Upsert File

### LINKED ISSUE #7: Add an action: Upsert S3 Object (aka File) By Key (aka ID/Filename)
[OIH Guideline](https://github.com/openintegrationhub/Connectors/blob/master/Adapters/AdapterBehaviorStandardization/StandardizedActionsAndTriggers.md#iteration-1-upsert-object-by-id)

## Config Fields
None Required

## Input Metadata
* Should Be Dynamic
* One input is the bucket name (enum, string, required).  This input should be an enum where the values are dynamically generated based on the buckets that can be written to given the AWS access key in the credentials.
* One input is the key name (string, required)
* One input is the Steward URL to upload (string, required)  **Note: It is not possible to access attachments from the component immediately prior to another component and attachment URLs can not be selected using the point and click link.  This is something that will be inconvienent at first but we will be able to over come it as the mapper is improved.**

## Behavior
[Some throwaway code that shows how to do this in JS](https://github.com/elasticio/Lloyds-s3-demo-component/blob/master/lib/actions/createS3Bucket.js)
Code should:
1. Read the data from the Steward URL.
2. Write the data to S3 using the bucket + key provided.  If the key already exists, allow AWS to overwrite the object.

## Output data
Emit an object where the body of the object contains all of the [Amazon headers in the resonse](https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPUT.html#RESTObjectPUT-responses)

## Gotchas
* [There are limits on the allowable chars in keys](https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingMetadata.html)  Perhaps we should throw an exception when we encounter a non-allowed key. 

## Not in scope (Limitations that can be done in future iterations)
* Setting Object Metadata in S3
* Support for objects that are so large, they can not fit in the memory of the component's docker container.
* Object more that 5 GB in size (so that they can always be uploaded in a single operation)
* Support for [Storage Classes](https://docs.aws.amazon.com/AmazonS3/latest/dev/storage-class-intro.html)
* Support for [Object Tagging](https://docs.aws.amazon.com/AmazonS3/latest/dev/object-tagging.html)
* Object compression (with zip, gzip, etc)
* Object encryption

### PR DESCRIPTION
**Pull request by jhorbulyk**
Resolves #7

### CODE REVIEW

File: component.json [hunk 1]
```diff
@@ -380,6 +381,16 @@
...
       }
+    },
+    "upsertFile": {
+      "main": "./lib/actions/upsertFile.js",
+      "title": "Write File to S3 From a Provided Attachment",
+      "help": {
+        "description": "Given a filename and a URL to an attachment stored in the platform, transfers the contents of the attachment to AWS S3. The component returns a summary of the written file. AWS S3 always overwrites the contents of the file if it already exists.",
+        "link": "/components/aws-s3/index.html#write-file-to-s3-from-a-provided-attachment"
+      },
+      "fields": {},
```
Comment (denyshld) [inline component.json:392]: No fields at all? What about bucket?

File: lib/actions/upsertFile.js [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+exports.process = async function (msg, cfg) {
+  const s3Client = new Client(this.logger, cfg);
+  const upsertObjectAction = new S3UpsertObject(this.logger, s3Client);
+  return upsertObjectAction.process(msg, cfg);
+};
+
+exports.getMetaModel = async function (cfg) {
+  const s3Client = new Client(this.logger, cfg);
+  const bucketNames = await s3Client.listBucketNames();
+  return {
```
Comment (denyshld) [inline lib/actions/upsertFile.js:53]: This metadata seems static, so better move it to component.json

Comment (jhorbulyk) [reply lib/actions/upsertFile.js:53]: It is static except for the enum of valid bucket names which is dynamic.

File: component.json (hunk 1 above)
Comment (jhorbulyk) [reply component.json:392]: Bucket exists as a field on the incoming message/metadata.  This way, if an integrator needs to dynamically route files to different buckets, they can do so. If they are only writing to a single bucket, they can select a valid enum member.

File: lib/actions/upsertFile.js (hunk 2 above)
Comment (denyshld) [reply lib/actions/upsertFile.js:53]: Ok, let's make it in this way

**Review by denyshld — APPROVED** [review]

**Pull request merged**